### ISSUE TITLE
Unable to fetch result from OAuth login

### ISSUE THREAD

**Issue by suyash515** [Web3Auth/Web3Auth#457]
I am getting the same issue on my end and these are the steps to reproduce the error:

1. Create a Custom Auth (Login Provider: Custom)
2. JWT Verifier ID: Sub
3. Set up the JWK Endpoint
4. JWT Validation: iss field set up
5. Set up a Plug and Play Project and using EVM based chain
6. Using the following code for the integration:

```
const web3authCore = new Web3AuthCore({
			chainConfig: { // this is ethereum chain config, change if other chain(Solana, Polygon)
				chainNamespace: CHAIN_NAMESPACES.EIP155,
				chainId: "0x1",
				rpcTarget: "mainnet_infura_url",
				blockExplorer: "https://etherscan.io/",
				ticker: "ETH",
				tickerName: "Ethereum"
			},
		});
		const adapter = new OpenloginAdapter({ adapterSettings: {
			network: "testnet",
			clientId: "client_id_from_web3auth_from_plugnplay",
			uxMode: "popup", // other option: popup
			loginConfig: {
				jwt: {
					name: "any name",
					verifier: "test",
					typeOfLogin: "jwt",
					clientId: "client_id_from_our_custom_login_provider",
				},
			},
		}});
		web3authCore.configureAdapter(adapter);
		await web3authCore.init();

		// call below code when user clicks on login button
		// it will use custom login with openlogin's authentication
		await web3authCore.connectTo(adapter.name, { 
			loginProvider: "jwt",
			extraLoginOptions: {
				domain: "redirect_uri_to_custom_login_provider",
				verifierIdField: "sub", 
				response_type: "token",
				scope:"email", // e.g. email openid
			}
		});
```

_Originally posted by @suyash515 in https://github.com/Web3Auth/Web3Auth/discussions/446#discussioncomment-3447463_

**Comment by rinzler99** [comment]
pls share console errors, if any

**Comment by suyash515** [comment]
![openlogin-console-error](https://user-images.githubusercontent.com/5890499/185980656-0e1bb307-b307-428f-9a60-8a94ff60b0d3.png)

@rinzler99 console errors.

**Comment by shahbaz17** [comment]
Hey @suyash515 

Which login provider are you using? This looks like the authentication didn't happen.

If you're, you are using a custom authentication provider(that are not natively supported on the verifier dashboard) that returns `idToken` after successful authentication. Then, simply use the below code to log in with Web3Auth:

```js
import { Web3AuthCore } from "@web3auth/core";
import { WALLET_ADAPTERS, CHAIN_NAMESPACES } from "@web3auth/base";

const web3auth = new Web3AuthCore({
  chainConfig: {
    chainNamespace: CHAIN_NAMESPACES.EIP155, // SOLANA, OTHER
    chainId: "0x1",
    rpcTarget: "https://rpc.ankr.com/eth",
    displayName: "Ethereum Mainnet",
    blockExplorer: "https://etherscan.io",
    ticker: "ETH",
    tickerName: "Ethereum",
  },
});

const openloginAdapter = new OpenloginAdapter({
  adapterSettings: {
    clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
    network: "testnet",
    uxMode: "redirect", // also support popup
    loginConfig: {
      jwt: {
        name: "Name of your choice",
        verifier: "YOUR-VERIFIER-NAME-ON-WEB3AUTH-DASHBOARD-FROM-CUSTOM-AUTH",
        typeOfLogin: "jwt",
        clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
      },
    },
  },
});

web3auth.configureAdapter(openloginAdapter);

await web3auth.init();

await web3auth.connectTo(WALLET_ADAPTERS.OPENLOGIN, {
  loginProvider: "jwt",
  extraLoginOptions: {
    id_token: idToken,
    verifierIdField: "sub", // same as your JWT Verifier ID
    domain: "https://YOUR-APPLICATION-DOMAIN" || "http://localhost:3000",
  },
});
```

Also, Please follow each step mentioned https://web3auth.io/docs/custom-authentication/byo-jwt-providers#set-up-custom-jwt-verifier for setting a Verifier.

**Comment by suyash515** [comment]
> Hey @suyash515
> 
> Which login provider are you using? This looks like the authentication didn't happen.
> 
> If you're, you are using a custom authentication provider(that are not natively supported on the verifier dashboard) that returns `idToken` after successful authentication. Then, simply use the below code to log in with Web3Auth:
> 
> ```js
> import { Web3AuthCore } from "@web3auth/core";
> import { WALLET_ADAPTERS, CHAIN_NAMESPACES } from "@web3auth/base";
> 
> const web3auth = new Web3AuthCore({
>   chainConfig: {
>     chainNamespace: CHAIN_NAMESPACES.EIP155, // SOLANA, OTHER
>     chainId: "0x1",
>     rpcTarget: "https://rpc.ankr.com/eth",
>     displayName: "Ethereum Mainnet",
>     blockExplorer: "https://etherscan.io",
>     ticker: "ETH",
>     tickerName: "Ethereum",
>   },
> });
> 
> const openloginAdapter = new OpenloginAdapter({
>   adapterSettings: {
>     clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
>     network: "testnet",
>     uxMode: "redirect", // also support popup
>     loginConfig: {
>       jwt: {
>         name: "Name of your choice",
>         verifier: "YOUR-VERIFIER-NAME-ON-WEB3AUTH-DASHBOARD-FROM-CUSTOM-AUTH",
>         typeOfLogin: "jwt",
>         clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
>       },
>     },
>   },
> });
> 
> web3auth.configureAdapter(openloginAdapter);
> 
> await web3auth.init();
> 
> await web3auth.connectTo(WALLET_ADAPTERS.OPENLOGIN, {
>   loginProvider: "jwt",
>   extraLoginOptions: {
>     id_token: idToken,
>     verifierIdField: "sub", // same as your JWT Verifier ID
>     domain: "https://YOUR-APPLICATION-DOMAIN" || "http://localhost:3000",
>   },
> });
> ```
> 
> Also, Please follow each step mentioned https://web3auth.io/docs/custom-authentication/byo-jwt-providers#set-up-custom-jwt-verifier for setting a Verifier.

@shahbaz17 
So, if I understand well, the custom login provider should return an additional parameter called idToken, right? What should the parameter contain?

The reason is that we have control over the custom login provider and can have any required values returned based on your requirements.

**Comment by shahbaz17** [comment]
Yes, that's correct. That should return an `idToken` in JWT format.

* JWT token header must contain the `kid` field.
* JWT token payload data must contain `iat` field.
* JWT token payload data must contain `iss` field. The value of this field will be used in JWT Validation field of Custom Auth Verifier on Web3Auth Dashboard..
* JWT token payload data must contain `aud` field. The value of this field will be used in JWT Validation field of Custom Auth Verifier on Web3Auth Dashboard.
* JWKS used to sign the token, should be exposed to Web3Auth. This endpoint URL to be used in the JWKS Endpoint field of Custom Auth Verifier on Web3Auth Dashboard.

Check https://web3auth.io/docs/custom-authentication/byo-jwt-providers#using-jsonwebtoken for more details on how to generate an idToken.

Using jose, it would look something like this:
```js
import * as jose from "jose";

import fs from "fs";
// openssl genrsa -out privateKey.pem 512
var privateKey = fs.readFileSync("privateKey.pem");
// openssl rsa -in privateKey.pem -pubout -out publicKey.pem
var publicKey = fs.readFileSync("publicKey.pem");
// https://my-authz-server/.well-known/jwks.json -> to be used in Custom Authentication as JWK Endpoint.

// Signing the JWT
const jwt = await new jose.SignJWT({ "urn:example:claim": true })
  .setProtectedHeader({ alg: "RS256", kid: "1bb9605c36e{your_kid}69386830202b2d" })
  .setIssuedAt()
  .setIssuer("https://my-authz-server")
  .setAudience("urn:my-resource-server")
  .setExpirationTime("2h")
  .sign(privateKey);
 ```

**Comment by suyash515** [comment]
@shahbaz17 Thank you. Let me try this and revert back.

**Comment by shahbaz17** [comment]
Hey @suyash515 

Have you tried this?

**Comment by suyash515** [comment]
@shahbaz17 We are still working on this. I will have an update on this later today.
